### PR TITLE
Test model logp before starting any MCMC chains

### LINKED ISSUE #4116: Test model logp before starting any MCMC chains
## Description of your problem
When a model contains `nan` observations, or observations that are outside of the support of the prior, the models logp evaluates to `-inf`.

Both `find_MAP` and `sample` don't check against this and fail in different ways that don't make the underlying problem obvious:
+ `pm.find_MAP` does two iterations, showing `-inf` for the logp, but it doesn't raise errors
+ `pm.sample` typically fails with `Bad initial energy`
+ `pm.sample` on Windows with multiprocessing crashes the child processes with uninformative error escalation (output similar as shown in #3403). If run with Jupyter notebooks, it even crashes the entire notebook server

Related issues:
enhancement #4107 would use the same information, but NOT raise

## Proposed solution
At the beginning of `find_MAP`/ `sample` and BEFORE creating child processes we should evaluate the model and raise an informative error.

Possibly even telling the user which variable was `nan`/`inf`?

### PR DESCRIPTION
**Pull request by StephenHogg**
This PR addresses [https://github.com/pymc-devs/pymc3/issues/4116](https://github.com/pymc-devs/pymc3/issues/4116) - making `find_MAP` and `sample` check their starting conditions before running any chains. I probably need to work out what the linting settings this repo uses are because it seems like a fair bit of formatting has changed.

### CODE REVIEW

**Comment by michaelosthege** [comment]
I'm sorry, but can you do the PR without applying `black` first? I appreciate the intent & it's fine to do in the end, but at this point it makes pinpointing the relevant changes close to impossible..

**Comment by MarcoGorelli** [comment]
> I'm sorry, but can you do the PR without applying `black` first? I appreciate the intent & it's fine to do in the end, but at this point it makes pinpointing the relevant changes close to impossible..

I think the opposite has happened - `black` has already been applied to the entire codebase, it looks like here they've reverted some of its changes - e.g., `black` wouldn't do this:
```diff
+ from .util import (chains_and_samples, dataset_to_point_dict, get_default_varnames, get_untransformed_name,
+                    is_transformed_name, update_start_vals)
```

@StephenHogg please see the [Python Style guide](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style) for this repo

**Comment by StephenHogg** [comment]
Sorry about this - had auto-linting on in my GUI and didn't realise. Have a look now, hopefully it's clearer.

File: pymc3/sampling.py [hunk 1]
```diff
@@ -419,6 +419,29 @@ def sample(
...
+    if start is None:
+        start = model.test_point
+    else:
+        if isinstance(start, dict):
+            update_start_vals(start, model.test_point, model)
+        else:
+            for chain_start_vals in start:
+                update_start_vals(chain_start_vals, model.test_point, model)
+
+    start_points = [start] if isinstance(start, dict) else start
```
Comment (michaelosthege) [inline pymc3/sampling.py]: If I remember correctly, the downstream code treats a `list` of `start` as "start points for each chain", which could explain your index error.

**Comment by StephenHogg** [comment]
yeah - the start points object doesn't get used further down, it's just for
checking the initial conditions in a way that makes handling both the case
when it's a dictionary and the case when it's an array easy. `Start` is
still what gets fed into functions later on, hence my confusion haha

**Comment by michaelosthege** [comment]
The error you posted in #4116 could also be a cause of invalid model test points. Could be that not all distributions have tests points.
Try to copy the model from the failing test case into a notebook to inspect it.

**Comment by StephenHogg** [comment]
Just for clarity - what's the path forward here? Sorry for the bother

File: pymc3/sampling.py [hunk 2]
```diff
@@ -419,6 +419,29 @@ def sample(
...
+    for elem in start_points:
+        if not set(elem.keys()).issubset(model.named_vars.keys()):
+            extra_keys = ", ".join(set(elem.keys()) - set(model.named_vars.keys()))
+            valid_keys = ", ".join(model.named_vars.keys())
+            raise KeyError("Some start parameters do not appear in the model!\n"
+                           "Valid keys are: {}, but {} was supplied".format(valid_keys, extra_keys))
+
+        initial_eval = model.check_test_point(test_point=elem)
+
+        if not np.all(np.isfinite(initial_eval)):
```
Comment (ColCarroll) [inline pymc3/sampling.py]: Feel free to ignore if you don't agree it is more readable.
```suggestion
    if np.any(~np.isfinite(initial_eval)):
```

File: pymc3/tuning/starting.py [hunk 3]
```diff
@@ -97,6 +97,14 @@ def find_MAP(
             "Valid keys are: {}, but {} was supplied".format(valid_keys, extra_keys)
         )
 
+    initial_eval = model.check_test_point(test_point=start)
+
+    if not np.all(np.isfinite(initial_eval)):
```
Comment (ColCarroll) [inline pymc3/tuning/starting.py]: Feel free to ignore if you don't agree it is more readable.
```suggestion
    if np.any(~np.isfinite(initial_eval)):
```

**Review by ColCarroll — COMMENTED** [review]
I think running this check is generally a good idea, but I think it needs to be put in at the right place. If you look at `sampling.py`, this change will get rid of a bunch of downstream code (there are some `if start is None`, but I don't think in a great way.

What if we had a `_check_start_point` function, and it got called at the end of `init_nuts`? I think it would contain `model.check_test_point`, and the nice error messages, but it would not do anything to the `start` argument passed to it.

**Comment by ColCarroll** [comment]
Sorry, I just read the attached issue, and it seems like that was steering @StephenHogg to put the changes where they are. Interested to hear if what you and @michaelosthege think!

**Comment by StephenHogg** [comment]
As a first time contributor I defer to Michael! :)

**Comment by michaelosthege** [comment]
> 
> 
> As a first time contributor I defer to Michael! :)

I think Colin is right: The block could easily become its own function. That also makes it easier to test, or improve.

**Comment by StephenHogg** [comment]
Ok - are you also saying the new function should be called at the end of `init_nuts` instead of where it is now?

**Comment by michaelosthege** [comment]
> 
> 
> Ok - are you also saying the new function should be called at the end of `init_nuts` instead of where it is now?

No, the NUTS initialization often suffers from `inf`/`NaN` and I think it's more useful to check that before initialization.

**Comment by ColCarroll** [comment]
The "main path" logic right now is:
- `start` goes into the function
- `init_nuts` is called, and comes up with its *own* `start_`
  - usually this is by taking the test point, and adding Uniform(-1, 1) noise
  - `start_` in `init_nuts` is used to initialize a mean for a running variance estimate
- `start_` is overwritten in the main loop whenever `start` is not `None`

I think Michael's right that `start_` also needs to be checked, since it is secretly being used internally (which is bad, but shouldn't be fixed here). But looking at the code, I think it can be done right after returning from `init_nuts`, otherwise there would be a ton of checks.

Concretely, I'm suggesting 
- a function that checks a user-supplied `start` value very early (around line 460, where there's already a check on `start`)
- again checking `start_` after it returns from `init_nuts` (around line 490)
- perhaps in a followup PR, having the function check other initialization schemes (there's an `else` branch for discrete models I've been ignoring...)

**Comment by StephenHogg** [comment]
@michaelosthege any more thoughts on the above? Would like to make sure I'm clear about what I'm coding up before starting again

**Comment by michaelosthege** [comment]
@StephenHogg listen to Colin on this one. He's much more literate in what the NUTS code is actually doing. With those checks in their own function, you can run them before & after NUTS initialization.
It costs a few model evaluations, but hey, this is MCMC sampling and it gives us more interpretable errors.

**Comment by StephenHogg** [comment]
I've shifted this into a function called `pm.util.check_start_vals`. In the process, I've spotted (and fixed) a bug in `model.check_test_point`. The function's `test_point` argument was being ignored at all times.

**Comment by StephenHogg** [comment]
Looking at the test output, it seems like a few other tests (e.g. `test_nuts_error_reporting` in `test_hmc.py`) are actually broken by the changes in this branch. Any guidance as to how to handle this appreciated.

**Comment by StephenHogg** [comment]
[Here's the output I get from pytest at this point](https://github.com/pymc-devs/pymc3/files/5540578/test.log), if that helps. Some of these are a bit mystifying, as I'm not sure why I'd be getting a max recursion depth error on a test that I've not touched, for instance. Will push one more change to format the error string a bit more nicely, but after that I think I'm probably stuck for now.

File: pymc3/util.py [hunk 4]
```diff
@@ -15,10 +15,11 @@
 import re
 import functools
 from typing import List, Dict, Tuple, Union
+import logging
 
 import xarray
 import arviz
-from numpy import asscalar, ndarray
+from numpy import asscalar, ndarray, isfinite, all
```
Comment (ColCarroll) [inline pymc3/util.py]: can you `import numpy as np`, or `from numpy import ..., all as np_all`? There is a python builtin named `all` that this will clobber.

**Review by ColCarroll — COMMENTED** [review]
this looks nice! I took a look at most of the the test failures, and they're surprisingly helpful. Feel free to ping again if you need more help, but I think this is close:
- Delete `test_hmc.py:test_nuts_error_reporting`. your check is a better one for the same behavior.
- `test_sampling.py:test_deterministic_of_observed` looks like a flake. let's ignore that and hope it goes away. if it doesn't, make the `rtol` bigger.
- `test_examples.py::TestLatentOccupancy::test_run` is interesting, and looks like a legit failure you found! In this case, the likelihood is passing parameters in the wrong order. It should be
`pm.ZeroInflatedPoisson("y", psi, theta, observed=y)` (note that `psi` and `theta` are switched). I imagine it was passing because the multipart sampling got everything to a reasonable place.
- Two failures in `pymc3/tests/test_step.py` can also be either deleted, or ported to the new exception you throw -- it looks like we have a `SamplingError` defined, which may be a good, specific error to raise instead of a `ValueError`.

File: pymc3/util.py (hunk 4 above)
Comment (canyon289) [reply pymc3/util.py]: Im being picky but please `import numpy as np`. Its so common to namespace under there that it would make this code easier to read if it follows the pattern

Comment (StephenHogg) [reply pymc3/util.py]: Have changed this, will push in a bit

**Comment by StephenHogg** [comment]
> this looks nice! I took a look at most of the the test failures, and they're surprisingly helpful. Feel free to ping again if you need more help, but I think this is close:
> 
> * Delete `test_hmc.py:test_nuts_error_reporting`. your check is a better one for the same behavior.

Done, thanks!

> * `test_sampling.py:test_deterministic_of_observed` looks like a flake. let's ignore that and hope it goes away. if it doesn't, make the `rtol` bigger.

Yeah I think it might be a flake, that whole file passed for me last time I looked. Updated: that whole file passed for me this time as well.

> * `test_examples.py::TestLatentOccupancy::test_run` is interesting, and looks like a legit failure you found! In this case, the likelihood is passing parameters in the wrong order. It should be
>   `pm.ZeroInflatedPoisson("y", psi, theta, observed=y)` (note that `psi` and `theta` are switched). I imagine it was passing because the multipart sampling got everything to a reasonable place.

Done, this passes now - thanks!

> * Two failures in `pymc3/tests/test_step.py` can also be either deleted, or ported to the new exception you throw -- it looks like we have a `SamplingError` defined, which may be a good, specific error to raise instead of a `ValueError`.

Ok, have changed over to returning a `SamplingError`. 

--------------------

I've noticed that `model.check_test_point` doesn't actually have test coverage (which is likely why that bug went unnoticed), so I'll add one for that. We're now down to about 20 or so cases that still need to be addressed in some way for all tests to pass. Aside from that, [here's the latest test log](https://github.com/pymc-devs/pymc3/files/5541864/test.log) and some areas that still bamboozle me a bit:

* `test_ode.py` I don't entirely follow what's going on here, I guess is the bottom line. Is there any reason why this error would be the case? `AttributeError: 'DifferentialEquation' object has no attribute '_get_test_value'`
* `test_distributions.py` This looks like a Theano problem, not sure how to diagnose what's going on here?
* `test_distributions_random.py` Not sure why I'm getting a `RecursionError` locally, given that I presume these tests should be expected to pass
* `test_minibatches.py` looks like it is failing on a line changed by the [recently-revealed fix to handle Theano casting](https://github.com/pymc-devs/pymc3/pull/4214#issuecomment-724930766). Not sure there's actually a relationship there seeing as there are a few chained operators involved, but mentioning for context:
```
pymc3/model.py:1800: in __init__
    self.tag.test_value = theano.compile.view_op(data).tag.test_value.astype(self.dtype)
```
In which case, what to do here? The `astype` thing seems like it's right but why this would cause a shape error is not clear to me
* `test_step.py` One of the failures here is because `pm.sample` is returning a `SamplingError` due to the new check rather than the `ParallelSamplingError` the test expects. We could change this simply enough, does it violate the spirit of the test though? Are all of these tests in fact needed, given we just got rid of one from `test_hmc.py` because it was kind of duplicating the tests for the new function?
* `test_types.py` The failures here are a bit cryptic: just "Chain failed".

**Comment by StephenHogg** [comment]
The only thing still failing at this point is one test in `test_step.py`, which is because it is expecting a `ParallelSamplingError` but the new check is just returning a `SamplingError`. I'm happy to shift this over, but wanted to check I'm not violating the spirit of this test?

Edit: the flaky test is also not passing, but that definitely passes locally

**Comment by StephenHogg** [comment]
Hi @ColCarroll - the only thing that still fails now is `test_sampling.py:test_deterministic_of_observed` when `FLOATX` is set to `float32` (doesn't appear to be a problem with 64 bit floats?). I can increase the rtol here but it's already `1e-3`, got any preference?

**Comment by ColCarroll** [comment]
This looks great! What if you loosen the tolerances on the test, but also open a bug and mention that it got worse when this PR was merged? That's very strange...

I think the last two things are:
- Remove draft status of this pr
- Add a line to RELEASE_NOTES.md

**Comment by StephenHogg** [comment]
Hi @ColCarroll changing the rtol to 1e-1 isn't helping unfortunately - looks like something is actually wrong here.

**Comment by StephenHogg** [comment]
Ok have set all tolerances in that one test to 1e-1 and all checks pass now

**Comment by michaelosthege** [comment]
An up to 10 % error in a deterministic addition? That sounds _very_ fishy.

The error seems to be related to the floatX setting. This could indicate a problem with a dtype - possibly of a test value, or induced by the new checking code?
Try to comment out the new checking code in the `sampling.py` and check if the test then passes with the old `rtol`.

Also try to print out the arrays that are being compared.

**Comment by ColCarroll** [comment]
Yeah, I'd hoped to be able to merge this without doing more work, since it seems like this isn't *your* bug, but Michael's right that this is pretty fishy. I'm going to try to also take a look at it this weekend, and will keep this thread up-to-date when I do!

**Comment by StephenHogg** [comment]
@ColCarroll subject to the outcome of #4246 I assume it's worth reverting the changes to `rtol` in this branch ahead of a rebase?

**Comment by ColCarroll** [comment]
Yeah, I would probably checkout the whole file. If the main pymc3 is called `upstream`, then it'd be 

```
git fetch upstream
git checkout upstream/master pymc3/tests/test_sampling.py
git rebase upstream/master
```

You'll also have to fix a merge conflict in `util.py` (it looks straightforward).

**Comment by twiecki** [comment]
Something went wrong with the rebase I'm afraid. You can also check out master, make a new branch, and cherry-pick your commits.

**Comment by StephenHogg** [comment]
Ok the rebase is not crazy now

**Comment by StephenHogg** [comment]
There seem to be more problems, unfortunately. The resolution of the merge conflict in `util.py` resulted in an unnecessary import, which makes linting throw an error. In addition, `ymc3/tests/test_examples.py::TestLatentOccupancy::test_run` throws an error, which has not happened before. Specifically, the new check is complaining there's a non-finite evaluation of logp. Here's the output:

```
E               pymc3.exceptions.SamplingError: Initial evaluation of model at starting point failed!
E               Starting values:
E               {'psi': array(0.5, dtype=float32), 'z': array([1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1,
E                      0, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0,
E                      0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0,
E                      0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
E                      1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0], dtype=int16), 'theta': array(5., dtype=float32), 'psi_logodds__': 0.0, 'theta_interval__': -2.944439}
E               
E               Initial evaluation results:
E               psi_logodds__       -1.390000
E               z                  -69.309998
E               theta_interval__    -3.050000
E               y                        -inf
E               Name: Log-probability of test_point, dtype: float64
```

I'll remove the unneeded import for now, but probably need guidance on the other error. @ColCarroll any thoughts?

**Comment by twiecki** [comment]
Seems like there are still conflicts though:

**Comment by StephenHogg** [comment]
> Seems like there are still conflicts though:

Yes, that's what I'm saying - I can either leave the conflict in, in which case I can't merge, or I can resolve the conflict in which case linting fails because there's an unneeded import. It's a Catch-22.

File: pymc3/tests/test_hmc.py [hunk 5]
```diff
@@ -57,16 +55,3 @@ def test_nuts_tuning():
 
     assert not step.tune
     assert np.all(trace["step_size"][5:] == trace["step_size"][5])
-
-
-def test_nuts_error_reporting(caplog):
```
Comment (twiecki) [inline pymc3/tests/test_hmc.py:62]: Why did you remove this test?

Comment (twiecki) [reply pymc3/tests/test_hmc.py:62]: Oh I get it, disregard.

File: pymc3/util.py [hunk 6]
```diff
@@ -188,6 +189,47 @@ def update_start_vals(a, b, model):
     a.update({k: v for k, v in b.items() if k not in a})
 
 
+def check_start_vals(start, model):
+    r"""Check that the starting values for MCMC do not cause the relevant log probability
+    to evaluate to something invalid (e.g. Inf or NaN)
+    Parameters
```
Comment (twiecki) [inline pymc3/util.py:196]:
```suggestion

    Parameters
```

**Comment by MarcoGorelli** [comment]
> > Seems like there are still conflicts though:
> 
> Yes, that's what I'm saying - I can either leave the conflict in, in which case I can't merge, or I can resolve the conflict in which case linting fails because there's an unneeded import. It's a Catch-22.

Shouldn't be a catch-22 :smile:

Can you try
```
git fetch --all --prune
git merge upstream/master
```
Then, in `pymc3/util.py`, you'll see something like
```
<<<<<<< HEAD
=======
from numpy import ndarray
>>>>>>> upstream/master
```
Change it to
```
```
(i.e., choose the current changes, ignore the incoming ones)

Then,
```
git add -u
git commit
git push -u origin HEAD
```

---

for more on `git`, I heartily recommend the [pro git book](https://git-scm.com/book/en/v2)

**Comment by StephenHogg** [comment]
As before - there's a mysterious new test failure

**Comment by ColCarroll** [comment]
Wow, CI got changed under you!

`test_examples.py::TestLatentOccupancy::test_run` got reverted during your merge, I think: `pm.ZeroInflatedPoisson("y", psi, theta, observed=y)` should be `pm.ZeroInflatedPoisson("y", psi, theta, observed=y)`

**Comment by StephenHogg** [comment]
This new error doesn't seem to have much to do with the code I wrote? Not sure, though

**Comment by MarcoGorelli** [comment]
> This new error doesn't seem to have much to do with the code I wrote? Not sure, though

can you check if that test passes when you run it locally?
```
pytest pymc3/tests/test_step.py::TestMLDA::test_acceptance_rate_against_coarseness
```

**Comment by StephenHogg** [comment]
@MarcoGorelli passes locally, had to update `theano-pymc` to version 1.0.11 to do it. Here's the output:

```
(pymc3) shogg@192:~/git/pymc3$ pytest pymc3/tests/test_step.py::TestMLDA::test_acceptance_rate_against_coarseness
================================================================================================================================== test session starts ==================================================================================================================================
platform darwin -- Python 3.7.5, pytest-5.0.1, py-1.9.0, pluggy-0.13.1
rootdir: /Users/shogg/git/pymc3, inifile: setup.cfg
collected 1 item                                                                                                                                                                                                                                                                        

pymc3/tests/test_step.py .                                                                                                                                                                                                                                                        [100%]

=================================================================================================================================== warnings summary ====================================================================================================================================
pymc3/tests/test_step.py::TestMLDA::test_acceptance_rate_against_coarseness
pymc3/tests/test_step.py::TestMLDA::test_acceptance_rate_against_coarseness
pymc3/tests/test_step.py::TestMLDA::test_acceptance_rate_against_coarseness
  /Users/shogg/git/pymc3/pymc3/step_methods/mlda.py:383: UserWarning: The MLDA implementation in PyMC3 is still immature. You should be particularly critical of its results.
    "The MLDA implementation in PyMC3 is still immature. You should be particularly critical of its results."

-- Docs: https://docs.pytest.org/en/latest/warnings.html
========================================================================================================================= 1 passed, 3 warnings in 13.26 seconds =========================================================================================================================
```

**Comment by StephenHogg** [comment]
Wait, all checks have passed now? Maybe the test was flaky?

**Comment by twiecki** [comment]
Yes, I think so. Thanks @StephenHogg!

**Comment by StephenHogg** [comment]
Whew, thanks

**Comment by ColCarroll** [comment]
+1 Thanks for sticking with us, @StephenHogg -- this was trickier than expected, but I think it will really improve lots of people's experiences.
